src/app/shell/personal-cabinet/shared-cabinet/messages/chat/chat.component.ts

```diff
@@ -104,34 +168,83 @@ export class ChatComponent implements OnInit, AfterViewInit, OnDestroy {
     );
   }
 
-  addListeners() {
+  private setChatSubscriptions(): void {
     this.chatRoom$
       .pipe(
         filter((chatRoom: ChatRoom) => !!chatRoom),
         takeUntil(this.destroy$)
       )
       .subscribe((chatRoom: ChatRoom) => {
         this.chatRoom = chatRoom;
-        this.store.dispatch(
-          new GetChatRoomMessages(this.chatRoom.id, this.messagesParameters)
-        );
+        this.store.dispatch(new GetChatRoomMessages(this.chatRoom.id, this.userRole, this.messagesParameters));
+        this.getChatMembersNames();
       });
+
     this.messages$
       .pipe(
         filter((messages: Message[]) => !!messages),
+        map((messages: Message[]) => messages.reverse()),
         takeUntil(this.destroy$)
       )
       .subscribe((messages: Message[]) => {
-        this.messages = messages;
+        let isUserDown =
+          this.chatEl.nativeElement.scrollTop === this.chatEl.nativeElement.scrollHeight - this.chatEl.nativeElement.clientHeight;
+
+        if (this.isAdditionalMessageLoading) {
+          this.messages = [...messages, ...this.messages];
+          this.isAdditionalMessageLoading = false;
+        } else {
+          this.messages.push(...messages);
+        }
+
+        if (isUserDown) {
+          this.scrollDown();
+        }
+      });
+  }
+
+  sendMessage(): void {
+    if (this.hubConnection.state === signalR.HubConnectionState.Connected && this.messageControl.value != '') {
```

Comment:
I'd prefer to do it as getter ---> this.hubConnection.state === signalR.HubConnectionState.Connected